File: 001-050/019/019.rb
class Problem019
   def is_leap_year(y)
      ret = false
      if y % 4 == 0
         if y % 100 != 0
            ret = true
         else
            if y % 400 == 0
               ret = true
            end
         end
      end
   end

   def count_months_in_which_first_day_is_sunday(number_of_days_in_month)
      count = 0
      days = 1 # 1901-01-01 is Tuesday
      p number_of_days_in_month

      1901.upto(2000) do |y|
         number_of_days_in_month.each do |month, num_of_days|
            days = days + num_of_days
            if is_leap_year(y) && month == :Feb
               days = days + 1
            end
            # Check the day of the week at the end of the month.
            # 6 is Saturday
            if days % 7 == 6
               count = count + 1
            end
         end
      end
      return count
   end

   def compute()
      number_of_days_in_month = {
         "Jan": 31,
         "Feb": 28,
         "Mar": 31,
         "Apr": 30,
         "May": 31,
         "Jun": 30,
         "Jul": 31,
         "Aug": 31,
         "Sep": 30,
         "Oct": 31,
         "Nov": 30,
         "Dec": 31,
      }   
      count_months_in_which_first_day_is_sunday(number_of_days_in_month)
   end

   #another solution
   require 'date'
   def count_months_in_which_first_day_is_sunday2(start_year, end_year)
      Date.new(start_year,1,1).upto(Date.new(end_year,12,31)).find_all { |d| d.mday == 1 && d.wday == 0 }.count
   end

   def compute_another_solution()
      count_months_in_which_first_day_is_sunday2(1901, 2000)
   end
end
p Problem019.new().compute()
p Problem019.new().compute_another_solution()

File: 001-050/019/019.rs

fn is_leap_year(y: u64) -> bool {
    if y % 400 == 0 {
        return true;
    }
    if y % 100 == 0 {
        return false;
    }
    if y % 4 == 0 {
        return true;
    }
    return false
}

fn days_of_year(y: u64) -> u64 {
    if is_leap_year(y) {
        return 366
    } else {
        return 365
    }
}

fn days_of_month(y: u64) -> [u64; 12] {
    [
        31,                                    // Jan
        if is_leap_year(y) { 29 } else { 28 }, // Feb
        31,                                    // Mar
        30,                                    // Apr
        31,                                    // May
        30,                                    // Jun
        31,                                    // Jul
        31,                                    // Aug
        30,                                    // Sep
        31,                                    // Oct
        30,                                    // Nov
        31,                                    // Dec
    ]
}

fn append_days(year: u64, offset: u64, result: &mut [u64; 7]) -> u64 {
    let mut day = offset;
    let days_of_month = days_of_month(year);
    for n in &days_of_month {
        result[day as usize] += 1;
        day = (day + *n) % 7;
    }
    return day
}

fn main() {
    let mut result = [0; 7];
    let mut day = 1; // Monday
    day = (day + days_of_year(1900)) % 7;
    for y in 1901..(2000 + 1) {
        day = append_days(y, day, &mut result);
        println!("{:?}", result)
    }
    println!("{}", result[0])
}